org.alloytools.alloy.classic.provider/src/main/java/org/alloytools/alloy/classic/solver/kodkod/AbstractKodkodSolver.java

```diff
@@ -0,0 +1,299 @@
+import edu.mit.csail.sdg.ast.Command;
+import edu.mit.csail.sdg.ast.ExprVar;
+import edu.mit.csail.sdg.ast.Sig;
+import edu.mit.csail.sdg.ast.Sig.Field;
+import edu.mit.csail.sdg.parser.CompModule;
+import edu.mit.csail.sdg.translator.A4Helper;
+import edu.mit.csail.sdg.translator.A4Options;
+import edu.mit.csail.sdg.translator.A4Options.SatSolver;
+import edu.mit.csail.sdg.translator.A4Solution;
+import edu.mit.csail.sdg.translator.A4Tuple;
+import edu.mit.csail.sdg.translator.A4TupleSet;
+import edu.mit.csail.sdg.translator.TranslateAlloyToKodkod;
+import kodkod.engine.satlab.SATFactory;
+
+public abstract class AbstractKodkodSolver extends AbstractSolver {
+
+	public AbstractKodkodSolver(Alloy core) {
+		super(core);
+	}
+
+	@Override
+	public Solution solve(TCommand command, SolverOptions optionsOrNull, Instance lowerBound) {
+		AbstractCommand c = (AbstractCommand) command;
+		return command(command.getModule(), optionsOrNull, c.getOriginalCommand());
+	}
+
+	private Solution command(Module module, SolverOptions optionsOrNull, Command command) {
+
+		SolverOptions options = super.processOptions(module, command, optionsOrNull);
+
+		CompModule orig = ((AlloyModuleClassic) module).getOriginalModule();
+		A4Reporter reporter = getReporter();
+		A4Options opt = new A4Options();
+		setOptions(opt, options);
+
+		Map<String, Atom> atoms = new HashMap<>();
+
+		A4Solution ai = getSolution(command, orig, reporter, opt);
+
+		TSig univ = module.getSig("univ")
+			.get();
+		// TSig Int = module.getSig("Int").get();
+		// TSig seqInt = module.getSig("seq/Int").get();
+
+		return new Solution() {
+
+			Atom createAtom(String o, TSig sig) {
+				return atoms.computeIfAbsent(o, k -> {
+					return new Atom(this, sig, o, o);
+				});
+			}
+
+			private Relation to(A4TupleSet set) {
+				List<IAtom> atoms = new ArrayList<>();
+
+				for (A4Tuple tuple : set) {
+					for (int j = 0; j < set.arity(); j++) {
+						String atomName = tuple.atom(j);
+						TSig sig = tuple.sig(j);
+						//
+						// A4 gives us the positive first
+						// integers with type "seq/Int". Not
+						// sure I understand why and it eats the
+						// symbols for integers. We create a
+						// special atom for the seq (not sure when
+						// and where it is needed) and then create
+						// the integer symbol.
+						//
+
+						if (sig.getName()
+							.startsWith("seq/Int")) {
+							createAtom("[" + atomName + "]", sig);
+							sig = module.getSig("Int")
+								.get();
+						}
+						Atom atom = createAtom(atomName, sig);
+						atoms.add(atom);
+					}
+				}
+
+				return new Relation(this, set.arity(), atoms);
+			}
+
+			Relation none = new Relation(this, 0, Collections.emptyList());
+
+			@Override
+			public IRelation none() {
+				return none;
+			}
+
+			@Override
+			public Iterator<Instance> iterator() {
+
+				if (!isSatisfied())
+					throw new IllegalStateException("This solution is not satisfied");
+
+				return new Iterator<Instance>() {
+
```

Comment:
this is way too much nesting.  I suggest extracting a class for Solution instead of using an anonymous class.

Reply:
Hmm, the problem is that the class refers to a lot of nicely local and private variables. When I turn it into a class I need to pass them as instance variables. This creates a lot of opportunities to screw up for no obvious reason.